### PR TITLE
Update typings based on new Typescript version.

### PR DESCRIPTION
**Pull request by dashmug**
Issue: https://github.com/date-fns/date-fns/issues/572

### CODE REVIEW

File: src/add_iso_years/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+declare module 'date-fns/add_iso_years' {
```
Comment (bradleyayers) [inline src/add_iso_years/index.d.ts]: You don't need the `declare module` syntax here. You should be able to simple have:

```
import {addISOYears} from 'date-fns'
export = addISOYears
```

File: src/locale/ar/index.d.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+declare module 'date-fns/locale/ar' {}
```
Comment (bradleyayers) [inline src/locale/ar/index.d.ts]: These locale files can simply be `export {}`

File: src/add_iso_years/index.d.ts (hunk 1 above)
Comment (dashmug) [reply src/add_iso_years/index.d.ts]: Thanks for the feedback. Do we only need the `declare module` syntax if the declaration file is not named `index.d.ts`?

File: scripts/generate_typings.js [hunk 3]
```diff
@@ -149,39 +133,46 @@ function getTypeScriptFnDefinition (fn) {
...
+  const name = fn.content.name
+  const snakeCaseName = fn.file.snakeCaseName
+  const filename = `./src/${snakeCaseName}/index.d.ts`
+  const moduleName = `date-fns/${snakeCaseName}`
 
-  const definition = `declare module 'date-fns/${name}' {}`
+  const typingString = [`declare module '${moduleName}' {`]
+    .concat(`  import {${name}} from 'date-fns'`)
+    .concat(`  export = ${name}`)
+    .concat('}')
```
Comment (leshakoss) [inline scripts/generate_typings.js]: Could you please add a new line at the end of the file? Like this: `.concat('}\n')`

File: scripts/generate_typings.js [hunk 4]
```diff
@@ -149,39 +133,46 @@ function getTypeScriptFnDefinition (fn) {
...
+
+  fns.forEach(generateTypeScriptFnTyping)
+}
+
+function generateTypeScriptLocaleTyping (locale) {
+  const snakeCaseName = locale.snakeCaseName
+  const name = `locale/${snakeCaseName}`
+  const filename = `./src/locale/${snakeCaseName}/index.d.ts`
+
+  const typingString = `declare module 'date-fns/${name}' {}`
```
Comment (leshakoss) [inline scripts/generate_typings.js]: Same here

File: package-lock.json [hunk 5]
```diff
@@ -0,0 +1,8223 @@
+{
```
Comment (leshakoss) [inline package-lock.json]: We use yarn for date-fns development, and the package doesn't have any non-dev dependencies, so this file is not really needed. Could you please delete it from the PR?

**Comment by dashmug** [comment]
@leshakoss @bradleyayers 
I'll test this more (including [@bradleyayers's feedback above](https://github.com/date-fns/date-fns/pull/573#discussion_r142312895)) later.

Thank you for the quick feedback.

File: src/add_iso_years/index.d.ts (hunk 1 above)
Comment (bradleyayers) [reply src/add_iso_years/index.d.ts]: It's only needed if it the declaration file is not for the adjacent JavaScript module.

**Comment by dashmug** [comment]
@leshakoss I have tested this on the following typescript versions and can confirm that it's working:
* v2.4.2
* v2.5.2
* v2.5.3 (latest)

Anything else I need to do to get this ready for merging?

**Comment by dashmug** [comment]
@leshakoss Wait, I take it back. I wasn't testing it right.

**Comment by timwangdev** [comment]
Any updates on this? @dashmugm , Thanks for the PR btw.

**Comment by ForbesLindesay** [comment]
My current workaround is a postinstall script of:

```js
const fs = require('fs');

const dirname = __dirname + '/../node_modules/date-fns';
const modules = fs
  .readdirSync(dirname)
  .filter(
    name =>
      fs.statSync(dirname + '/' + name).isDirectory() &&
      /^[a-z0-9]+(\_[a-z0-9]+)*$/.test(name),
  );

modules.forEach(name => {
  const camelName = name.replace(/_(.)/g, (_, c) => c.toUpperCase());
  fs.writeFileSync(
    dirname + '/' + name + '/index.d.ts',
    [
      'import {' + camelName + "} from 'date-fns';",
      'export = ' + camelName + ';',
    ].join('\n'),
  );
});
```

**Comment by kossnocorp** [comment]
Thank you a lot! I just finished backporting build scripts from v2 and going to ship this update.

**Comment by kossnocorp** [comment]
Sorry for fake news, but I had to revert the merge. See: https://github.com/date-fns/date-fns/issues/572#issuecomment-335781855

If anyone here wants to help, please chime in!

**Comment by timwangdev** [comment]
@kossnocorp 
@ForbesLindesay 's postinstall script works on my production project. We could just run the script once and call it a commit.

**Comment by kossnocorp** [comment]
@timwangdev this is great!

Am I properly understand that the working solution is to introduce `index.d.ts` next to every function with such content?

```js
import { functionName } from 'date-fns';
export = functionName;
```

**Comment by timwangdev** [comment]
@kossnocorp Yep!

**Comment by kossnocorp** [comment]
Nice! Then I'll update the scripts and release a new version.

**Comment by cbengtson85** [comment]
@kossnocorp 
Just started migrating to date-fns from moment and running into this issue.  Is there an ETA when the new version will be released?  Thanks!

**Comment by ForbesLindesay** [comment]
@kossnocorp  any chance you could push a release soon?

**Pull request merged**